### ISSUE TITLE
DateTime.TimeOfDay  conversion

### ISSUE THREAD

**Issue by sergonfire** [linq2db/linq2db#1982]
This is my simplified request:
```cs
from rec in db.Table1 where 
   (from c in db.Table2 where c.BegTime != null ? c.BegTime < rec.ResTime.TimeOfDay : true).Any() select rec;
```
c.BegTime has type TimeSpan,  rec.ResTime has type DateTime.

Framework created this compare fragment :
``` SQL
[c_1].[BegTime] < Cast(Convert(Char, [rec_4].[p_1], 114) as DateTime)
```
and generated SQL exception during execute:
The data types time and datetime are incompatible in the less than operator.

My opinion, conversion must has: `Cast(Convert(Char, [rec_4].[p_1], 114) as Time)`
or `Cast(rec_4].[p_1], 114) as Time)`.

### Environment details
linq2db version: 2.9.4
Database Server:  MS SQL
Database Provider: System.Data.SqlClient
Operating system: Windows10
.NET Framework: 4.8

**Comment by sergonfire** [comment]
sorry,
from rec in db.Table1 where
(from c in db.Table2 where c.BegTime != null ? c.BegTime < rec.ResTime.TimeOfDay : true select c).Any() select rec;

**Comment by MaceWindu** [comment]
Will see if we can fix it for 2.9.5, otherwise it will go to 3.0, but looks like it is doable. Also should work if ResTime declared as `DateTimeOffset`, but probably it is not an option for you :)

**Comment by sergonfire** [comment]
Thank you, now work comparison via 2 DateTime objects:

var req =from rec in db.Table1 where
(from c in db.Table2 where c.BegTime != null ? new DateTime(rec.ResTime.Year,rec.ResTime.Month,rec.ResTime.Day,c.BegTime.Hours,c.BegTime.Minutes,0)  < rec.ResTime : true select c).Any() select rec;
